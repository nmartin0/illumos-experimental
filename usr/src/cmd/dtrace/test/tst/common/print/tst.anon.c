/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 */

/*
 * Copyright 2025 Oxide Computer Company
 */

/*
 * This embeds type information with anonymous structures and unions and passes
 * it to a function which allows the pid provider to get at it with various
 * forms of the print() action in tst.anon.ksh.
 */

#include <stdio.h>
#include <string.h>

struct elves {
	int feanor;
	struct {
		int fingolfin;
	};
	union {
		int maedhros;
		int fingon;
		struct {
			int aredhel;
			union {
				int turgon;
				struct {
					int tuor;
					int idril;
					struct {
						union {
							int earendil;
							int elwing;
						};
						int silmaril;
					};
				};
				union {
					int maeglin;
					int morgoth;
					union {
						int balrog;
						int gondolin;
						int glorfindel;
						int ecthelion;
					};
				};
			};
		};
	};
	struct {
		int elrond;
		int elros;
	};
};

void
mandos(struct elves *elves)
{
	(void) fprintf(stderr, "%x\n", elves->feanor);
}

int
main(void)
{
	struct elves elves;
	(void) memset(&elves, 0, sizeof (elves));
	elves.feanor = 0x1497;
	elves.fingolfin = 0x467;
	elves.maedhros = 0x587;
	elves.turgon = 0x510;
	elves.idril = 0x525;
	elves.earendil = 0x7777;
	elves.silmaril = 0x9999;
	elves.elrond = 0x3021;
	elves.elros = 0x442;

	mandos(&elves);
};
