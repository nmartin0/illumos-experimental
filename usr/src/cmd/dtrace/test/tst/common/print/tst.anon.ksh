#!/usr/bin/ksh
#
# This file and its contents are supplied under the terms of the
# Common Development and Distribution License ("CDDL"), version 1.0.
# You may only use this file in accordance with the terms of version
# 1.0 of the CDDL.
#
# A full copy of the text of the CDDL should have accompanied this
# source.  A copy of the CDDL is also available via the Internet at
# http://www.illumos.org/license/CDDL.
#

#
# Copyright 2025 Oxide Computer Company
#

#
# Verify that we can reasonably print and walk various anonymous unions and
# structs.
#

if (( $# != 1 )); then
        printf "%s\n" "expected one argument: <dtrace-path>" >&2
        exit 2
fi

dtrace=$1
$dtrace -c ./tst.anon.exe -qs /dev/stdin <<EOF
pid\$target::mandos:entry
{
	print(*args[0]);
	printf("\n");
	print(args[0]->turgon);
	print(args[0]->balrog);
	print(args[0]->elrond);
	print(args[0]->silmaril);
	print(*(userland struct pid\`elves *)arg0);
	printf("\n");
	printf("feanor: 0x%x\n", args[0]->feanor);
	printf("fingolfin: 0x%x\n", args[0]->fingolfin);
	printf("maedhros: 0x%x\n", args[0]->maedhros);
	printf("aredhel: 0x%x\n", args[0]->maedhros);
	printf("fingon: 0x%x\n", args[0]->fingon);
	printf("turgon: 0x%x\n", args[0]->turgon);
	printf("tuor: 0x%x\n", args[0]->tuor);
	printf("idril: 0x%x\n", args[0]->idril);
	printf("earendil: 0x%x\n", args[0]->earendil);
	printf("elwing: 0x%x\n", args[0]->elwing);
	printf("silamril: 0x%x\n", args[0]->silmaril);
	printf("maeglin: 0x%x\n", args[0]->maeglin);
	printf("morgoth: 0x%x\n", args[0]->morgoth);
	printf("balrog: 0x%x\n", args[0]->balrog);
	printf("gondolin: 0x%x\n", args[0]->gondolin);
	printf("glorfindel: 0x%x\n", args[0]->glorfindel);
	printf("elrond: 0x%x\n", args[0]->elrond);
	printf("elros: 0x%x\n", args[0]->elros);
	exit(0);
}
EOF
