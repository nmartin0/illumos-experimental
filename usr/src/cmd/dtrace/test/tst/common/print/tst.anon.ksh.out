struct elves {
    int feanor = 0x1497
    struct  {
        int fingolfin = 0x467
    }
    union  {
        int maedhros = 0x587
        int fingon = 0x587
        struct  {
            int aredhel = 0x587
            union  {
                int turgon = 0x510
                struct  {
                    int tuor = 0x510
                    int idril = 0x525
                    struct  {
                        union  {
                            int earendil = 0x7777
                            int elwing = 0x7777
                        }
                        int silmaril = 0x9999
                    }
                }
                union  {
                    int maeglin = 0x510
                    int morgoth = 0x510
                    union  {
                        int balrog = 0x510
                        int gondolin = 0x510
                        int glorfindel = 0x510
                        int ecthelion = 0x510
                    }
                }
            }
        }
    }
    struct  {
        int elrond = 0x3021
        int elros = 0x442
    }
}
int 0x510
int 0x510
int 0x3021
int 0x9999
struct elves {
    int feanor = 0x1497
    struct  {
        int fingolfin = 0x467
    }
    union  {
        int maedhros = 0x587
        int fingon = 0x587
        struct  {
            int aredhel = 0x587
            union  {
                int turgon = 0x510
                struct  {
                    int tuor = 0x510
                    int idril = 0x525
                    struct  {
                        union  {
                            int earendil = 0x7777
                            int elwing = 0x7777
                        }
                        int silmaril = 0x9999
                    }
                }
                union  {
                    int maeglin = 0x510
                    int morgoth = 0x510
                    union  {
                        int balrog = 0x510
                        int gondolin = 0x510
                        int glorfindel = 0x510
                        int ecthelion = 0x510
                    }
                }
            }
        }
    }
    struct  {
        int elrond = 0x3021
        int elros = 0x442
    }
}
feanor: 0x1497
fingolfin: 0x467
maedhros: 0x587
aredhel: 0x587
fingon: 0x587
turgon: 0x510
tuor: 0x510
idril: 0x525
earendil: 0x7777
elwing: 0x7777
silamril: 0x9999
maeglin: 0x510
morgoth: 0x510
balrog: 0x510
gondolin: 0x510
glorfindel: 0x510
elrond: 0x3021
elros: 0x442

